lib/AST/ASTPrinter.cpp

```diff
@@ -2373,6 +2392,143 @@ void PrintAST::printExtension(ExtensionDecl *decl) {
   }
 }
 
+/// Functions to determine which features a particular declaration uses. The
+/// usesFeatureNNN functions correspond to the features in Features.def.
+
+static bool usesFeatureStaticAssert(Decl *decl) {
+  return false;
+}
+
+static bool usesFeatureAsyncAwait(Decl *decl) {
+  if (auto func = dyn_cast<AbstractFunctionDecl>(decl)) {
+    if (func->hasAsync())
+      return true;
+  }
+
+  return false;
+}
+
+static bool usesFeatureMarkerProtocol(Decl *decl) {
+  if (auto proto = dyn_cast<ProtocolDecl>(decl)) {
+    if (proto->isMarkerProtocol())
+      return true;
+  }
+
+  if (auto ext = dyn_cast<ExtensionDecl>(decl)) {
+    for (const auto &req: ext->getGenericRequirements()) {
+      if (req.getKind() == RequirementKind::Conformance &&
+          req.getSecondType()->castTo<ProtocolType>()->getDecl()
+              ->isMarkerProtocol())
+        return true;
+    }
+
+    for (const auto &inherited : ext->getInherited()) {
+      if (auto inheritedType = inherited.getType()) {
+        if (inheritedType->isExistentialType()) {
+          auto layout = inheritedType->getExistentialLayout();
+          for (ProtocolType *protoTy : layout.getProtocols()) {
+            if (protoTy->getDecl()->isMarkerProtocol())
+              return true;
+          }
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+static bool usesFeatureActors(Decl *decl) {
+  if (auto classDecl = dyn_cast<ClassDecl>(decl)) {
+    if (classDecl->isActor())
+      return true;
+  }
+
+  if (auto ext = dyn_cast<ExtensionDecl>(decl)) {
+    if (auto classDecl = ext->getSelfClassDecl())
+      if (classDecl->isActor())
+        return true;
+  }
+
+  return false;
+}
```

Comment:
Unless I'm mistaken, all three of these probably ought to walk into any parameter or result types and look for the feature being used there, too. For instance, if you write:

```
public func runAsync(_ fn: () async -> Void) { ... }
```

That probably ought to be guarded by `#if $AsyncAwait`, even though `runAsync(_:)` itself isn't an async function. Similarly:

```
public func transact(with account: BankAccount)
````

Probably ought to be guarded by `#if $Actors` if `BankAccount` is an actor class, even though `transact(with:)` isn't itself an actor class.

Reply:
Good point!

Reply:
I went ahead and did a bunch of this checking. It is not comprehensive, I'm sure, although we could probably do some additional meta programming to make it less boilerplate-y and more comprehensive.